Update README with notes about PVData/normative types

amittal94 opened the pull request:
This PR is to update flatbuffers documentation. It corresponds to issue #15 .

amittal94 committed:
update documentation with pva0 and implementation overview

amittal94 commented:
I made a start on this, but need help with how to approach it and what to cover.

btspiers commented on README.md:
```diff
@@ -51,7 +54,107 @@ decoded_value = deserialise_f142(buf)
+also encoded in the byte string which is used during the deserialisation.
+
+### ADAr
+This is intended for image data from channel access with limited metadata.
```
This should say AreaDetector I think, rather than channel access. ADPluginKafka bypasses EPICS to send the `ADAr` buffers directly into kafka

btspiers commented on README.md:
```diff
@@ -51,7 +54,107 @@ decoded_value = deserialise_f142(buf)
+- timestamp: nanoseconds past epoch
+- dimensions: an array containing details on the dimensions of the image
+- data_type: type of the data stored in the array
+- data: elements of the array
```
Might be worth mentioning that this is a raw byte array, not an array of `data_type` values

amittal94 replied:
I did not actually know this...as the raw bytes were identical to pixel values in the examples I had

btspiers commented on README.md:
```diff
@@ -51,7 +54,107 @@ decoded_value = deserialise_f142(buf)
+on [normative types][normative-types]) for `NTScalarAny`, `NTNDArray` and `NTTable`.
+
+#### Implementation Details
+This schema was built based on [normative types][normative-types]. A `PVData` object is defined
```
Might be worth saying something like "the EPICS v4 normative types specification" rather than just "normative types"?

btspiers commented on README.md:
```diff
@@ -51,7 +54,107 @@ decoded_value = deserialise_f142(buf)
+which contains the following fields:
+- data: the value of the PV as sent from pv access, in the form of one of the supported normative
+types
+- source_name: a string value identifying source
```
Identifying how? I think this is intended to be a PV name? Not sure if that's always applicable though, or what it should be if that's not applicable

amittal94 replied:
As per Stephen, we might have automated analysis programs running and producing data into Kafka. No actual PV name in that case.

stephen-jd-dann replied:
Ben has a good point, that a bit more detail would be good here. For example:

```suggestion
- source_name: the source of the data; for example, the name of a PV
```

When we add other data sources I'd like to add another field identifying the type of the data source. For now, we're focussing on PVs, but with an eye to future-proofing (which means calling this `source_name` instead of `pv_name`).

btspiers commented on README.md:
```diff
@@ -51,7 +54,107 @@ decoded_value = deserialise_f142(buf)
+- source_name: a string value identifying source
+
+Currently `NTScalarAny` is used to handle both `NTScalar` and `NTScalarArray` types of data. `NTTable` is
+also supported for any potential use cases.
```
Maybe mention NTNDArray as well? In particular that it's used for AreaDetector/image data

btspiers commented on README.md:
```diff
@@ -155,3 +258,4 @@ generation process.
 
 [streaming-data-types]: https://github.com/ess-dmsc/streaming-data-types
 [python-streaming-data-types]: https://github.com/ess-dmsc/python-streaming-data-types
+[normative-types]: https://docs.epics-controls.org/en/latest/pv-access/Normative-Types-Specification
```
This link didn't work for me without the `.html` on the end
```suggestion
[normative-types]: https://docs.epics-controls.org/en/latest/pv-access/Normative-Types-Specification.html
```

btspiers reviewed (commented):
Looks good overall I think - there's some details I'd like to additionally include in places. Stephen may well have more/better insights than me, as I think he's more familiar with the code at hand

One thing I was wondering but without a natural place to comment so I'll just put it here: do we have a settled NT representation for waveforms? These docs say that pva0 is planned to supersede the `f142`, `wa00` and `ADAr` formats, and I'd say it's clear to me how it's going to supersede `f142` and `ADAr` but I'm not really sure how it's indended to supersede `wa00` yet

amittal94 commented:
> One thing I was wondering but without a natural place to comment so I'll just put it here: do we have a settled NT representation for waveforms? These docs say that pva0 is planned to supersede the `f142`, `wa00` and `ADAr` formats, and I'd say it's clear to me how it's going to supersede `f142` and `ADAr` but I'm not really sure how it's indended to supersede `wa00` yet

Yes I should make a note of that, we do not have clarity on waveforms yet, as we were not sure how we will handle it, so that will be done later. This is not quite a completed version yet, hence I mention that deprecated in the future.

amittal94 committed:
documentation improvements for accuracy

amittal94 committed:
improve source name description

stephen-jd-dann commented on README.md:
```diff
@@ -69,7 +172,7 @@ any code using the Python [`streaming_data_types`][python-streaming-data-types]
 with little effort beyond changing the import paths.
 
-### Major differences with upstream
+### Major differences with upstream for f142 and ADAr
```
You can probably remove this header as well as the next paragraph.

stephen-jd-dann commented on README.md:
```diff
@@ -51,7 +54,107 @@ decoded_value = deserialise_f142(buf)
+- value: main value of the pv in question, which can be any non string scalar
+- timestamp: nanoseconds past epoch
+- status: an enum representing alarm_status, details are in the schema
+- schema: an enum representing alarm_severity, details are in the schema
```
```suggestion
- severity: an enum representing alarm_severity, details are in the schema
```

stephen-jd-dann commented on README.md:
```diff
@@ -51,7 +54,107 @@ decoded_value = deserialise_f142(buf)
+### f142
+This is intended for scalar and array values from channel access with limited metadata.
+
+#### Implementation Details
```
For `f142`, `ADAr`, and `wa00`, there's not much more detail here than what's in the schema file. Which isn't really surprising if you used that as your main source.

stephen-jd-dann commented on README.md:
```diff
@@ -51,7 +54,107 @@ decoded_value = deserialise_f142(buf)
+Note that each of these will have different arguments, in particular for `pva0` a PVData object
+as defined in data_types.py is expected.
+
+## Currently supported schemas and details
```
I think this entire section would benefit from being more of a user guide or brief explanation than a technical reference.

stephen-jd-dann reviewed (commented):
I think we're starting to run into the limits of what you can put in a readme, and we need a proper documentation solution.

In the mean-time, I think we'd be better off with a "user guide" style of writing, with examples, to cover `pva0` serialisation and deserialisation, along with `ca_to_pva`. Rather than trying to explain what all the fields mean.

amittal94 commented:
Very well, I will try to restructure it into a user guide style, and we can revisit it then.

amittal94 committed:
take a user guide centered approach to README

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
+    alarm_severity=AlarmSeverity.NO_ALARM,
+)
+
+decoded_value = deserialise_f142(encoded_value)
```
Might be nice to show field access here.

amittal94 replied:
you meant on the derserialised data correct?

stephen-jd-dann replied:
Yes, exactly. Sure, you're showing how to deserialise data, but you're giving the user an opaque object. They need to know how to look inside it.

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
+
+This is intended for scalar and array values from channel access.
 
 #### Implementation Details
```
I'd reorder this a bit. I'd go straight into the example and follow that with important technical notes. Headers probably not needed.

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
-
-As value can take multiple data types it is implemented as a Value union, made of tables of the
+
+As the value can take multiple data types it is implemented as a Value union, made of tables of the
```
After reading this, I want to know which data types are supported and how to specify a data type when serialising. Some more (small) examples might be useful.

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
-This kind of data is serialised using the `serialise_ADAr` function, which takes each field as an
-argument.
-The bytes can be deserialised using the corresponding `deserialise_ADAr`.
+This is intended for image data from AreaDetector via ADPluginKafka. While direct serialisation is
+possible, ADPluginKafka is the currently preffered usage.
```
Excellent summary. You could add a link to the ADPluginKafka repo.

stephen-jd-dann commented on README.md:
````diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
+#### Usage (Python)
+
+```python
+from epac.flatbuffers.area_detectot_ADAr import (
````
Typo?

```suggestion
from epac.flatbuffers.area_detector_ADAr import (
```

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
+    serialise_ADAr,
+)
+
+encoded_value = serialise_f142(42,
```
```suggestion
encoded_value = serialise_ADAr(42,
```

We don't have a tool to check these snippets work sadly.

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
 
-#### Implementation Details
-wa00 is defined in the schema as a structure with the following fields:
+This is intended for waveform data from channel access via two PVs.
```
I think you need to define "waveform data" more clearly. My usage is a little unusual. Something like:

    A *waveform* is made of two arrays of the same length, one with x co-ordinates and the other
    with y co-ordinates. Examples include an optical spectrum and an oscilloscope trace.

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
+    values_y_array=np.array([6, 7, 8, 9, 10, 6], dtype=np.uint64),
+    timestamp=datetime.now(tz=timezone.utc),
+    x_timestamp=datetime.now(tz=timezone.utc) - timedelta(minutes=5),
+    y_unit="vs",
+    x_unit="ss",
```
Nitpick:

```suggestion
    y_unit="V",
    x_unit="s",
```

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
+            ],
+)
+
+decoded_value = deserialise_ADAr(encoded_value)
```
Again, a couple of examples of field access would be good.

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
+    x_unit="ss",
+)
 
+decoded_value = deserialise_wa00(encoded_value)
```
Again, some examples of field access.

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
 - source_name: the source of the data; for example, the name of a PV
 
 Currently `NTScalarAny` is used to handle both `NTScalar` and `NTScalarArray` types of data. `NTNDArray` is used for AreaDetector/Image data.
 NTTable` is also supported for any potential use cases.
```
```suggestion
`NTTable` is also supported for any potential use cases.
```

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
+
+This is intended for various types of data from pv access.
 
 #### Implementation Details
```
I'd remove this header and rephrase slightly. E.g.

    This schema was built based on the EPICS V4 [normative types][normative-types]. A `PVData` object
    contains the value of the PV as sent from pv access, in the form of one of the supported data types,
    as well as some additional metadata, such as the name of the PV (the `source_name`).

stephen-jd-dann commented on README.md:
```diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
 - Enums such as display form are not handled via EnumT as per the normative types specification, rather
 they use the style of flatbuffer enums.
 - To convert between the case of Alarm enums, and the case of display form, they pydantic objects come
 with parser support, which formats the data as required.
```
This may be too technical. It's what's at the top of your mind, because of all the work you did developing it, but these details are not of interest to the user (unless perhaps they're very familiar with normative types). I'd go with a simple:

    Enums are provided for the alarm status and severity, as well as the display format. These are
    automatically converted from the corresponding parts of the normative types.

amittal94 replied:
A bias I did not consider, will simplify it.

stephen-jd-dann commented on README.md:
````diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
 - To convert between the case of Alarm enums, and the case of display form, they pydantic objects come
 with parser support, which formats the data as required.
 
+```python
````
This example is too big and doesn't solve typical user problems. Let's split it up into smaller examples and put a bit of body text before each one explaining what we're doing.

Here's what I'd go with:
- Use `p4p` to get some data and serialise it (shows typical usage, as well as any special `p4p` options needed)
- Serialise a minimal dict; leave out most of the fields (short; shows that data can be left out)
- Same, but with the pydantic types

stephen-jd-dann commented on README.md:
````diff
@@ -59,101 +57,263 @@ as defined in data_types.py is expected.
 
+#### Usage (Python)
+
+```python
+from epac.flatbuffers.ca_to_pva import CaToNtConverter
+from epac.flatbuffers import data_types as dt
+value = {
+    "value": 42.5,
+    "pvname": "some pv name",
+    "status": 1,
+    "precision": 2,
+    "units": "V",
+    "severity": 0,
+    "timestamp": 1713201234.567,
+    "upper_disp_limit": 100.0,
+    "lower_disp_limit": 0.0,
+    "upper_ctrl_limit": 95.0,
+    "lower_ctrl_limit": 5.0
+}
+ca_scalar_obj = dt.CAScalarAny(**value)
+# value could also have been passed directly to convert_scalar
+nt_scalar_object = CaToNtConverter().convert_scalar(ca_scalar_obj)
+pv_data_object = PVData(data=nt_scalar_object, sourceName=self.pv_name)
+```
````
This example would probably be improved by showing how it's used with `pyepics`. It's intended to be a callback, right? Does it work with `PV.get()`?

stephen-jd-dann commented on schemas/wa00_waveform.fbs:
```diff
@@ -11,7 +11,7 @@ table WaveFormArray{
     y_data_type: DType;         // The type of the data stored in the y_data array
     x_data: [ubyte] (required);  // Elements in the x array
-    y_data: [ubyte] (required);  // Elements in the x array
+    y_data: [ubyte] (required);  // Elements in the y array
```
Nice spot.

stephen-jd-dann reviewed (changes requested):
Much nicer, thanks. I think there's still space to move things around to make it friendlier, but the biggest improvement would come from splitting up the `pva0` example to show more explicitly how to solve common problems.

Might also be good to make `pva0` the first entry, since it's supposed to be the default going forwards.

amittal94 committed:
expand readme examples and simplify them for clarity

amittal94 committed:
move pva0 up

stephen-jd-dann commented on README.md:
```diff
@@ -41,15 +41,14 @@ from epac.flatbuffers.logdata_f142 import (
 
-decoded_value = deserialise_f142(encoded_value)
-print(decoded_value.value)
+deserialised_data = deserialise_f142(serialised_data)
```
I think by removing `print(deserialised_data.value)` you've lost something from this example.

amittal94 replied:
Ah I see what you mean, there is a an example of attribute access in the pva0 example which I intended to put be put in the first example, but forgot this example came earlier.

stephen-jd-dann commented on README.md:
````diff
@@ -59,165 +58,75 @@ as defined in data_types.py is expected.
-
-decoded_value = deserialise_wa00(encoded_value)
-```
+can be referred to for most cases with minimal changes. Refer to [epac-forwarder][epac-forwarder] to see
````
```suggestion
can be referred to for most cases with minimal changes. Refer to [epac-forwarder] to see
```

should work IIUC

stephen-jd-dann commented on README.md:
````diff
@@ -59,165 +58,75 @@ as defined in data_types.py is expected.
-
-decoded_value = deserialise_wa00(encoded_value)
-```
+can be referred to for most cases with minimal changes. Refer to [epac-forwarder][epac-forwarder] to see
````
Not a big fan of referring to another project for examples. It's real code rather than example code so it's not going to be super clear to a novice. They'll also need to navigate an unfamiliar project to find the right code.

Also, currently this project is public and that one is private.

OTOH, I don't think this code is worth the space in the README file, so maybe we could have a directory of examples?

amittal94 replied:
yeah an example directory makes sense, I thought the README would get too long if I tried to cover too much.

stephen-jd-dann commented on README.md:
```diff
@@ -59,165 +58,75 @@ as defined in data_types.py is expected.
+from epac.flatbuffers.data_types import PVData, NTScalarAny, NTNDArray, NTTable
+from epac.flatbuffers.pva0_data import serialise_data, deserialise_data
+
+context = Context("pva", nt=False)
```
I'd add a comment explaining what `nt=False` does here and how important it is.

stephen-jd-dann commented on README.md:
```diff
@@ -59,165 +58,75 @@ as defined in data_types.py is expected.
+    data=NTScalarAny(**value), sourceName=pv_name
+)
+serialised_data = serialise_data(pv_data_object)
+deserialised_data = deserialise_data(serialised_data)
```
You can probably leave this out of most of the examples - you only really need it once.

stephen-jd-dann commented on README.md:
```diff
@@ -59,165 +58,75 @@ as defined in data_types.py is expected.
+deserialised_data = deserialise_data(serialised_data)
+
+
+# the defined pydantic types can also freely be used
```
Looking at it now I'm not sure this example is very useful.

stephen-jd-dann commented on README.md:
```diff
@@ -314,6 +184,136 @@ nt_scalar_object = CaToNtConverter().convert_scalar(ca_scalar_obj)
+
+pv = epics.PV(
+    "EPAC-DEV:CAM1:stats1:Net_RBV",
+    form="ctrl",
```
It's probably worth calling this out with a comment

stephen-jd-dann reviewed (dismissed):
This is looking a lot nicer now.

The pva0 and f142 examples are more involved than the others, and you may want to preface them with a short explanations of what they're doing.

I'm not keen on referring to `epac-forwarder` so much. It's better than having nothing, but don't trick yourself into thinking they count as examples. If you found a library on the internet and you wanted to learn it and its examples were "look at this other project; no, we won't tell you where in the code" how helpful would you find that?

I'd say it's fine for now, and we may never find time to do something better, but it's not a replacement for proper examples.

amittal94 committed:
Documentation improvements:
- removed unnecessary alias for link
- reduced repetition in examples
- clearer comments
- removed unnecessary example

amittal94 commented:
I think this good to go based on your recommendations bar removing the usage of epac-forwarder and referencing it for further examples. Should I remove that and create an examples directory? Or leave that in a separate issue and PR?

stephen-jd-dann reviewed (dismissed):
When a PR has 49 comments it is not a good idea to suggest increasing the scope.

In other words, an examples directory belongs in a separate PR.

amittal94 committed:
comment on nt=False

amittal94 commented:
That may be on me. Just added a single comment about the nt=False, please re-approve.

stephen-jd-dann reviewed (approved):
Hey. Hey. Hey. It's not your fault that I'm picky.
